### PR TITLE
Updated progress bar for asynchronous queries

### PR DESCRIPTION
**Pull request by MSebanc**
# Description

Improved progress bar handling of asynchronous query calls

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/main/connection.cpp [hunk 1]
```diff
@@ -71,6 +76,12 @@ std::unique_ptr<QueryResult> Connection::executeWithParams(PreparedStatement* pr
     return clientContext->executeWithParams(preparedStatement, std::move(inputParams));
 }
 
+std::unique_ptr<QueryResult> Connection::executeWithParamsWithId(
```
Comment (andyfengHKU) [inline src/main/connection.cpp:79]: Do we need this? I'm pretty sure if we prepare, we cannot support multiple statements. You can add a test case on this.

File: src/include/common/task_system/progress_bar.h [hunk 2]
```diff
@@ -23,30 +23,28 @@ class ProgressBar {
 
     void addPipeline();
 
-    void finishPipeline();
+    void finishPipeline(std::string id);
```
Comment (andyfengHKU) [inline src/include/common/task_system/progress_bar.h:26]: better to rename to queryID and comment when this is needed.

File: src/include/main/connection.h [hunk 3]
```diff
@@ -74,6 +76,10 @@ class Connection {
      */
     KUZU_API std::unique_ptr<QueryResult> executeWithParams(PreparedStatement* preparedStatement,
         std::unordered_map<std::string, std::unique_ptr<common::Value>> inputParams);
+
+    std::unique_ptr<QueryResult> executeWithParamsWithId(PreparedStatement* preparedStatement,
```
Comment (andyfengHKU) [inline src/include/main/connection.h:80]: I would remove this.

File: src/include/processor/execution_context.h [hunk 4]
```diff
@@ -1,17 +1,26 @@
...
 
 struct ExecutionContext {
+    std::string queryId;
     common::Profiler* profiler;
     main::ClientContext* clientContext;
 
-    ExecutionContext(common::Profiler* profiler, main::ClientContext* clientContext)
-        : profiler{profiler}, clientContext{clientContext} {}
+    ExecutionContext(common::Profiler* profiler, main::ClientContext* clientContext, std::string id)
+        : profiler{profiler}, clientContext{clientContext} {
```
Comment (andyfengHKU) [inline src/include/processor/execution_context.h:16]: I think @ray6080 will need a system level query id soon. So let's make this id generated from database which you will need a counter and a lock. Then ExecutionContext should always be created with an id and does not need to generate one in the constructor.


Also I would suggest using a integer based id.

File: src/main/client_context.cpp [hunk 5]
```diff
@@ -482,7 +483,7 @@ std::unique_ptr<QueryResult> ClientContext::executeAndAutoCommitIfNecessaryNoLoc
     }
     auto queryResult = std::make_unique<QueryResult>(preparedStatement->preparedSummary);
     auto profiler = std::make_unique<Profiler>();
-    auto executionContext = std::make_unique<ExecutionContext>(profiler.get(), this);
+    auto executionContext = std::make_unique<ExecutionContext>(profiler.get(), this, id);
```
Comment (andyfengHKU) [inline src/main/client_context.cpp:486]: So here try to get a system level id from database.

File: src/main/client_context.cpp [hunk 6]
```diff
@@ -503,7 +504,7 @@ std::unique_ptr<QueryResult> ClientContext::executeAndAutoCommitIfNecessaryNoLoc
         }
     } catch (std::exception& e) {
         transactionContext->rollback();
-        progressBar->endProgress();
+        progressBar->endProgress(executionContext->queryId);
```
Comment (andyfengHKU) [inline src/main/client_context.cpp:507]: We prefer ID over Id as a convention I think.

File: tools/nodejs_api/src_cpp/include/node_connection.h [hunk 7]
```diff
@@ -78,16 +79,21 @@ class ConnectionExecuteAsyncWorker : public Napi::AsyncWorker {
     ~ConnectionExecuteAsyncWorker() override = default;
 
     void Execute() override {
+        std::string id = common::UUID::toString(
```
Comment (andyfengHKU) [inline tools/nodejs_api/src_cpp/include/node_connection.h:82]: ditto here try to get a system level id from database.

File: src/include/main/client_context.h [hunk 8]
```diff
@@ -109,8 +109,9 @@ class KUZU_API ClientContext {
     // Query.
     std::unique_ptr<PreparedStatement> prepare(std::string_view query);
     std::unique_ptr<QueryResult> executeWithParams(PreparedStatement* preparedStatement,
-        std::unordered_map<std::string, std::unique_ptr<common::Value>> inputParams);
-    std::unique_ptr<QueryResult> query(std::string_view queryStatement);
+        std::unordered_map<std::string, std::unique_ptr<common::Value>> inputParams,
```
Comment (andyfengHKU) [inline src/include/main/client_context.h:112]: ditto I would not change this API given we don't have multi query in a prepared statement.

**Review by andyfengHKU — CHANGES_REQUESTED** [review]
Overall LGTM but I need to take another look for system level query id.

Also @mewim should take a look on the js part.

File: src/include/common/task_system/progress_bar_display.h [hunk 9]
```diff
@@ -14,11 +16,13 @@ class ProgressBarDisplay {
 
     virtual ~ProgressBarDisplay() = default;
 
-    // Update the progress of the pipeline and the number of finished pipelines
-    virtual void updateProgress(double newPipelineProgress, uint32_t newNumPipelinesFinished) = 0;
+    // Update the progress of the pipeline and the number of finished pipelines. queryID is used to
+    // identify the query
```
Comment (andyfengHKU) [inline src/include/common/task_system/progress_bar_display.h:18]: Say queryID is needed when we track progress of multiple queries asynchronously.

File: src/include/processor/execution_context.h [hunk 10]
```diff
@@ -1,17 +1,20 @@
 #pragma once
 
 #include "common/profiler.h"
+#include "common/types/uuid.h"
```
Comment (andyfengHKU) [inline src/include/processor/execution_context.h]: this should be no longer needed.

**Comment by mewim** [comment]
I do not think we should print anything from the Node.js API side. If it is for debug only, you should delete the printing.

File: src/include/main/database.h [hunk 11]
```diff
@@ -139,6 +142,8 @@ class Database {
     std::unique_ptr<extension::ExtensionOptions> extensionOptions;
     std::unique_ptr<DatabaseManager> databaseManager;
     common::case_insensitive_map_t<std::unique_ptr<storage::StorageExtension>> storageExtensions;
+    uint64_t queryID;
```
Comment (andyfengHKU) [inline src/include/main/database.h:145]: Let's wrap this into a struct called QueryIDGenerator

File: src/include/main/database.h [hunk 12]
```diff
@@ -121,6 +122,8 @@ class Database {
     common::case_insensitive_map_t<std::unique_ptr<storage::StorageExtension>>&
     getStorageExtensions();
 
+    uint64_t getNextQueryID();
```
Comment (andyfengHKU) [inline src/include/main/database.h:125]: ditto move to QueryIDGenerator.

File: tools/nodejs_api/src_cpp/include/node_progress_bar_display.h [hunk 13]
```diff
@@ -1,6 +1,10 @@
 #pragma once
 
+#include <optional>
+#include <unordered_set>
+
 #include "common/task_system/progress_bar_display.h"
+#include "common/task_system/terminal_progress_bar_display.h"
```
Comment (andyfengHKU) [inline tools/nodejs_api/src_cpp/include/node_progress_bar_display.h:7]: include terminal_progress_bar does not sound correct to me.

File: tools/nodejs_api/src_cpp/node_connection.cpp [hunk 14]
```diff
@@ -41,6 +41,8 @@ Napi::Value NodeConnection::InitAsync(const Napi::CallbackInfo& info) {
 
 void NodeConnection::InitCppConnection() {
     this->connection = std::make_shared<Connection>(database.get());
+    connection->getClientContext()->getProgressBar()->setDisplay(
+        std::make_shared<NodeProgressBarDisplay>());
```
Comment (andyfengHKU) [inline tools/nodejs_api/src_cpp/node_connection.cpp:45]: not related to this PR. But I think you might be able to use unique_ptr for display in ProgressBar class.

**Review by andyfengHKU — APPROVED** [review]

File: tools/nodejs_api/src_cpp/include/node_connection.h [hunk 15]
```diff
@@ -2,6 +2,7 @@
 
 #include <utility>
 
+#include "common/types/uuid.h"
```
Comment (andyfengHKU) [inline tools/nodejs_api/src_cpp/include/node_connection.h:5]: remove this

**Pull request merged**